src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,720 @@
+
+    // Records count for debug
+    uint32_t i;
+
+    bool done;
+    bool same_input;
+    bool is_switch_wal;
+} WalFilterState;
+
+/***********************************************************************************************************************************
+Render as string for logging
+***********************************************************************************************************************************/
+static void
+walFilterToLog(const WalFilterState *const this, StringStatic *const debugLog)
+{
+    strStcFmt(
+        debugLog,
+        "{record_num: %u, step: %u is_begin: %s, page_offset: %zu, input_offset: %zu, rec_buf_size: %zu,"
+        " header_offset: %zu, got_len: %zu, tot_len: %zu}",
+        this->i,
+        this->currentStep,
+        this->is_begin ? "true" : "false",
+        this->page_offset,
+        this->input_offset,
+        this->rec_buf_size,
+        this->header_offset,
+        this->got_len,
+        this->tot_len
+        );
+}
+
+#define FUNCTION_LOG_WAL_FILTER_TYPE                                                                                               \
+    ZstDecompress *
+#define FUNCTION_LOG_WAL_FILTER_FORMAT(value, buffer, bufferSize)                                                                  \
+    FUNCTION_LOG_OBJECT_FORMAT(value, walFilterToLog, buffer, bufferSize)
+
+static inline
+void
+checkOutputSize(Buffer *const output, const size_t size)
+{
+    if (bufRemains(output) <= size)
+    {
+        bufResize(output, bufUsed(output) + size);
+    }
+}
+
+// Reading the next page from the input buffer. If the input buffer is exhausted, remember the current step and returns true.
+// In this case, we should exit the process function to get a new input buffer.
+static inline
+bool
+readPage(WalFilterState *const this, const Buffer *const input, const ReadStep step)
+{
+    if (this->input_offset >= bufUsed(input))
+    {
+        ASSERT(step != noStep);
+        this->currentStep = step;
+        this->input_offset = 0;
+        this->same_input = false;
+        return false;
+    }
+    this->page = bufPtrConst(input) + this->input_offset;
+    this->input_offset += this->walPageSize;
+    this->page_offset = 0;
+    this->currentStep = noStep;
+    this->current_header = (XLogPageHeaderData *) this->page;
+    this->page_offset += XLogPageHeaderSize(this->current_header);
+
+    // Make sure that WAL belongs to supported Postgres version, since magic value is different in different versions.
+    if (this->current_header->xlp_magic != this->walInterface->header_magic)
+    {
+        THROW_FMT(FormatError, "%s - wrong page magic", strZ(pgLsnToStr(this->recPtr)));
+    }
+
+    lstAdd(this->headers, this->current_header);
+
+    return true;
+}
+
+static inline uint32_t
+getRecordSize(const unsigned char *const buffer)
+{
+    return ((XLogRecord *) (buffer))->xl_tot_len;
+}
+
+static bool
+readRecord(WalFilterState *const this, const Buffer *const input)
```

Comment:
I'm a bit confused: what does the return value of this function mean?

Reply:
Okay, why in the `readBeginOfRecord` and `getEndOfRecord` functions do we handle when the `readRecord` function returns false, that is, "new input buffer is needed to continue reading"?